### PR TITLE
refactor main.py and environment variables

### LINKED ISSUE #4: get_country_code feature
- client input country name --> 3 letter country code

### PR DESCRIPTION
**Pull request by Isaiah-Ang**

### CODE REVIEW

File: api/codes.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    return convert_city(city, country_code)
+
+
+def convert_city(city, country):
+    city_code = str()
+    for x in city_data:
+        if x['name_translations']['en'] == city and x['country_code'] == country:
+            # print(x['name_translations']['en'])
+            # city_code.append(x['name_translations']["en"])
+            city_code = x['code']
```
Comment (tiongg) [inline api/codes.py]: Personally, I think its better to write [declarative code over imperative code](https://www.linkedin.com/pulse/imperative-vs-declarative-programming-javascript-yehuda-margolis/), since it is overall neater and more readable. The following code here can be expressed as 
```
city_index = index(city_data, lambda x: x['name_translations']['en'] == city and x['country_code'] == country)
city_code = city_data[city_index]['code]
```

File: api/codes.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    # print(len(city_data))
+
+
+def convert_airport_code(city):
+    airport_code = list()
+    for x in airport_data:
+        if x['city_code'] == city:
+            airport_code.append(
+                {"name": x['name_translations']['en'], "code": x['code']})
+
```
Comment (tiongg) [inline api/codes.py]: Consider using [list comprehension](https://www.w3schools.com/python/python_lists_comprehension.asp) here
```
airport_code = [{"name": x['name_translations']['en'], "code": x['code']} for x in airport_data if x['city_code'] == city]

File: api/codes.py (hunk 1 above)
Comment (Isaiah-Ang) [reply api/codes.py]: I feel imperative is better for me as it is easier to understand and goes through step by step on how the code will work.

Comment (tiongg) [reply api/codes.py]: Ok, atleast put a `break` so that it ends the loop after it finds

Comment (tiongg) [reply api/codes.py]: Also whut happens if the city is not found?

Comment (Isaiah-Ang) [reply api/codes.py]: What package did the index method come from?

Comment (Isaiah-Ang) [reply api/codes.py]: > Also whut happens if the city is not found?

The function will return an empty variable. I will need to cater for that before moving it onto the next function.

Comment (tiongg) [reply api/codes.py]: > 

Sry I'm high, supposed to be `city_data.index(...)`

Comment (Isaiah-Ang) [reply api/codes.py]: > > 
> 
> Sry I'm high, supposed to be `city_data.index(...)`

i don't think lambda works with index... just tried it and it throws exceptions. Ill see if there are any other functions or methods i can use.

Comment (tiongg) [reply api/codes.py]: Ah, try
```
def find(l, f):
     return next((l[i] for i in xrange(len(l)) if f(l[i])), None)

city = find(city_data, lambda x: x['name_translations']['en'] == city and x['country_code'] == country)
city_code = city['code']
```

Comment (tiongg) [reply api/codes.py]: The `find` function can be put into a seperate utils file for use else where. Kinda odd tho, expected python to have a built in list find function

Comment (Isaiah-Ang) [reply api/codes.py]: > Ah, try
> 
> ```
> def find(l, f):
>      return next((l[i] for i in xrange(len(l)) if f(l[i])), None)
> 
> city = find(city_data, lambda x: x['name_translations']['en'] == city and x['country_code'] == country)
> city_code = city['code']
> ```

oh i managed to get it to work already using 2 different methods

either:
`city_index = list(filter(
        lambda x: city_data[x]['name_translations']['en'] == city and city_data[x]['country_code'] == country, range(len(city_data))))`

or

`city_index = [city_data.index(x) for x in city_data if x['name_translations']['en'] == city and x['country_code'] == country]`

Comment (Isaiah-Ang) [reply api/codes.py]: > The `find` function can be put into a seperate utils file for use else where. Kinda odd tho, expected python to have a built in list find function

I would guess the find function would be the filter function for python... not too sure tho hahaha

**Comment by tiongg** [comment]
Remember to include in readme where the data is sourced from. (Should update after https://github.com/Isaiah-Ang/AirTix_Telebot/pull/11 is merged into main tho)

File: api/retrieve_data.py [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
+load_dotenv()
 
 
 def get_airports_json():
     airport_json_url = 'https://travelpayouts-travelpayouts-flight-data-v1.p.rapidapi.com/data/en-GB/airports.json'
     header = {
-        'X-Access-Token': '9ae0f41c70f0955ad052f47fa2c20c32',
-        'X-RapidAPI-Key': '7531220fafmsh31047231a74586ap14d55bjsn27c675ca9aea',
+        'X-Access-Token': os.getenv('X_ACCESS_API'),
+        'X-RapidAPI-Key': os.getenv('X_RAPIDAPI_KEY'),
```
Comment (tiongg) [inline api/retrieve_data.py:13]: Env should be something like `TRAVEL_PAYOUTS_RAID_API` or something. Don't think `X_ACCESS_API` is very descriptive

File: main.py [hunk 4]
```diff
@@ -13,104 +13,207 @@
...
+    details['origin_city'] = text
+
+    airport = codes.convert_country_code(
+        details['origin_city'], details['origin_country'])
+
+    buttons = list()
+
+    for x in airport:
+        buttons.append([InlineKeyboardButton(
+            f"{x['name']} ({x['code']})", callback_data=f"{x['code']}/{str(SELECT_ORIGIN_AIRPORT)}")])
```
Comment (tiongg) [inline main.py]: Can consider using list comprehension

File: main.py [hunk 5]
```diff
@@ -13,104 +13,207 @@
...
+    return SELECT_ORIGIN_AIRPORT
+
+# async def origin_airport_button(update: Update, context: ContextTypes.DEFAULT_TYPE):
+
+
+async def destination_country(update: Update, context: ContextTypes.DEFAULT_TYPE):
+    query = update.callback_query
+
+    await query.answer()
+    details['origin_airport'] = (query.data).split('/')[0]
```
Comment (tiongg) [inline main.py:86]: Whuts the split for? Can leave a comment

File: main.py [hunk 6]
```diff
@@ -13,104 +13,207 @@
...
+    print(query.data)
```
Comment (tiongg) [inline main.py]: remove

File: main.py [hunk 7]
```diff
@@ -13,104 +13,207 @@
...
+    details['destination_city'] = text
+
+    airport = codes.convert_country_code(
+        details['destination_city'], details['destination_country'])
+
+    buttons = list()
+
+    for x in airport:
+        buttons.append([InlineKeyboardButton(
+            f"{x['name']} ({x['code']})", callback_data=f"{x['code']}/{str(SELECT_DESTINATION_AIRPORT)}")])
```
Comment (tiongg) [inline main.py]: Consider comprehension or a map call

**Comment by tiongg** [comment]
Just curious, how will multiple users using the bot work? I believe the data in `details` obj will be overriden when another user messages the bot

**Comment by Isaiah-Ang** [comment]
> Just curious, how will multiple users using the bot work? I believe the data in `details` obj will be overriden when another user messages the bot

For this, I will have to save the session for the chat between each user into the database. I will create a new issue for this to be worked on in the future.

File: api/codes.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+def convert_country_code(city, country):
+    country_code = cc.convert(names=country, to='ISO2')
+    return convert_city(city, country_code)
+
+
+def convert_city(city, country):
+    city_index = list(filter(lambda x: city_data[x]['name_translations']['en'] ==
+                      city and city_data[x]['country_code'] == country, range(len(city_data))))
+    city_code = city_data[city_index[0]]['code']
```
Comment (tiongg) [inline api/codes.py]: hmm, this piece of code still looks funny to me...maybe add the `[0]` at the end of the list() since we are declaring that as `city_index`, rather than a `city_list`. Also, if possible, extract the `list(filter(lambda))` combo into a utils file, so we can reuse it elsewhere

File: api/codes.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+    # Check if the city exists
+
+    return convert_airport_code(city_code)
+
+
+def convert_airport_code(city):
+    airport_code = [{"name": x['name_translations']['en'], "code": x['code']}
+                    for x in airport_data if x['city_code'] == city]
+    return airport_code
```
Comment (tiongg) [inline api/codes.py]: Consider typing the input args and return values.
https://peps.python.org/pep-0484/
https://docs.python.org/3/library/dataclasses.html
tldr;
```
from dataclasses import dataclass

@dataclass
class AirportData:
    name: str
    code: str

def convert_airport_code(city: str) -> [AirportData]:
    ...
```

File: api/codes.py (hunk 8 above)
Comment (Isaiah-Ang) [reply api/codes.py]: i will make the utils file in the future if I do really need to use it again.

File: api/codes.py [hunk 10]
```diff
@@ -17,15 +22,18 @@ def convert_country_code(city, country):
...
 
 
-def convert_airport_code(city):
+def convert_airport_code(city) -> AirportData:
     airport_code = [{"name": x['name_translations']['en'], "code": x['code']}
                     for x in airport_data if x['city_code'] == city]
     return airport_code
+
+
+convert_country_code('Melbourne', 'Australia')
```
Comment (tiongg) [inline api/codes.py]: Remove this. If you wanna do adhoc testing on functions, create a `testing.py` file, add it to `.gitignore`, then import the functions and run it there instead

**Comment by tiongg** [comment]
LGTM!
